### PR TITLE
Add RNN for batch normalization

### PR DESCRIPTION
**Pull request by agibsonccc**
## What changes were proposed in this pull request?
Add RNN batch norm support
(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

## Quick checklist

The following checklist helps ensure your PR is complete:

- [ X] Eclipse Contributor Agreement signed, and signed commits - see [IP Requirements](https://deeplearning4j.org/eclipse-contributors) page for details
- [X ] Reviewed the [Contributing Guidelines](https://github.com/eclipse/deeplearning4j/blob/master/CONTRIBUTING.md) and followed the steps within.
- [ X] Created tests for any significant new code additions.
- [X ] Relevant tests for your changes are passing.

### CODE REVIEW

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java [hunk 1]
```diff
@@ -390,10 +390,10 @@ public INDArray preOutput(INDArray x, TrainingMode training, LayerWorkspaceMgr w
         int dim = 1;
         if(x.rank() == 4 && layerConf().getCnn2DFormat() == CNN2DFormat.NHWC)
             dim = 3;
-        if(x.size(dim) != layerConf().getNOut()){
+  /*      if(x.size(dim) != layerConf().getNOut()) {
             throw new IllegalArgumentException("input.size(" + dim + ") does not match expected input size of " + layerConf().getNIn()
                     + " - got input array with shape " + Arrays.toString(x.shape()));
-        }
+        }*/
```
Comment (treo) [inline deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java]: This looks like it is temporary, did you forget to revert that? If not remove it outright instead of commenting it out.

File: nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/nativeblas/Nd4jCpu.java [hunk 2]
```diff
@@ -12994,6 +12994,7 @@ public native void scatterUpdate(@Cast("Nd4jPointer*") PointerPointer extraPoint
 // #include <ops/declarable/headers/BarnesHutTsne.h>
 // #include <ops/declarable/headers/images.h>
 // #include <ops/declarable/headers/updaters.h>
+// #include <ops/declarable/headers/decoder.h>
```
Comment (treo) [inline nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/nativeblas/Nd4jCpu.java:12997]: Ideally changes to Nd4Cpu.java should be included in the original Pull Request that adds the C++ ops.

Comment (agibsonccc) [reply nd4j/nd4j-backends/nd4j-backend-impls/nd4j-native/src/main/java/org/nd4j/nativeblas/Nd4jCpu.java:12997]: @treo Nd4jCpu.java is actually generated from nd4j-cpu-presets. This should actually just be on .gitignore.

File: deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java (hunk 1 above)
Comment (agibsonccc) [reply deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/layers/normalization/BatchNormalization.java]: @treo yeah that's why it's not marked for review yet :)

**Review by treo — APPROVED** [review]

**Comment by joel-a** [comment]
Thank you for taking care of this. When will these changes be reflected in the Maven snapshot?

**Comment by MPdaedalus** [comment]
thanks for the update

**Comment by agibsonccc** [comment]
@joel-a  snapshots should be in there now. I've also ensure we have automation built in  for publishing snapshots every other day to get around ossrh auto purging snapshots every 3 days. We  will also be doing a service release here within the next few weeks to address some of the other changes the community found from doing the post github actions migration. Let us know if you have any other issues.

**Comment by joel-a** [comment]
@agibsonccc  I'm seeing a new error that I believe is related to this issue. It seems like the batch normalization layer is not outputting the correct type. Here's the error message:

```
Exception in thread "main" java.lang.IllegalStateException: Invalid input for 1D CNN layer (layer name = "conv1d_16"): expect RNN input type with size > 0. Got: InputTypeFeedForward(10,NWC)
	at org.deeplearning4j.nn.conf.layers.Convolution1DLayer.setNIn(Convolution1DLayer.java:104)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration.getLayerActivationTypes(ComputationGraphConfiguration.java:554)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration.addPreProcessors(ComputationGraphConfiguration.java:458)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraphConfiguration(KerasModel.java:613)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:635)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:624)
	at org.deeplearning4j.nn.modelimport.keras.KerasModelImport.importKerasModelAndWeights(KerasModelImport.java:168)
```

Adding an activation layer to the model after the batch normalization layer triggers this error:

```
[batchNormError2.zip](https://github.com/eclipse/deeplearning4j/files/6638978/batchNormError2.zip)


in_layer = Input((25, 25,))
x = Conv1D(filters=10, kernel_size=2)(in_layer)
x = BatchNormalization()(x)
x = Activation('relu')(x)
out_layer = Conv1D(512, 3, strides=1,
                          dilation_rate=3,
                          padding='same',
                          activation='linear')(x)
model = Model(in_layer, out_layer)
model.save("batchNormError2.h5")
```

DL4j code to import model:
```
String modelPath = "batchNormError2.h5";
ComputationGraph model = KerasModelImport.importKerasModelAndWeights(modelPath);
```
 I'm including this simple test model here for convenience:

[batchNormError2.zip](https://github.com/eclipse/deeplearning4j/files/6638981/batchNormError2.zip)

Thank you very much for the great work and attention to this issue.

**Comment by agibsonccc** [comment]
@joel-a https://github.com/eclipse/deeplearning4j/pull/9365
